File: Python/pkg/component/message.py
"""
    Package 'component' provides definitions for various objects meeting the definition of a protocol component.
"""

import json

from enum import Enum

class DataType(Enum):
    INT16 = "int16"
    INT32 = "int32"
    FLOAT = "float"
    STRING = "string"
    BIT = "bit"

"""
    Objects of class Message represent a specific message within a protocol.

	Required fields: name.
	Other fields are stored in a dictionary corresponding with protocol elements.

"""

class Message:

    def __init__(self, config: json, logger):
        # Setup logger
        self.logger = logger
        # Ensure all required fields are present
        if 'name' in config:
            # Initialize message fields
            self.msg_dict = {}
            self.message_byte_array = None
            # Create a dictionary representation of the message  
            for item in config:
                self.msg_dict[item] = config[item]
            self.__initialize()
        else:
            raise AttributeError("Message: name field is required")
    
    def __initialize(self):
        for key in self.msg_dict:
            setattr(self, key, self.msg_dict[key])
              
    def log(self):
        """
        Logs message information.

        Args:
            None
        Returns:
            None
        Raises:
            None
        """
        self.logger.debug(f"Message Name: {self.name}")
        for key in self.msg_dict:
            self.logger.debug(f"Message Dictionary Item: {key} Value: {self.msg_dict[key]}")

    def set_message(self, msg: bytearray):
        """
        Sets the message bytearray, which will be used for the duration of the message.

        Args:
            msg (bytearray): The message bytearray.
        Returns:
            None.
        Raises:
            None.
        """
        if self.message_byte_array is None and msg is not None:
            self.message_byte_array = msg

    def get_data_byte_count(self):
        """
        Safely calculates the byte count for a given message.

        Args:
            None.
        Returns:
            int: The calculated byte count.
        Raises:
            None.
        """
        length = self.get_data_length()
        data_type = DataType.INT16.value
        if self.msg_dict.get("data_type") is not None:
            data_type = self.msg_dict["data_type"].lower()
        # Match data type
        if data_type == DataType.INT16.value:
            return length * 2
        if data_type == DataType.INT32.value:
            return length * 4
        if data_type == DataType.FLOAT.value:
            return length * 4
        if data_type == DataType.STRING.value:
            return length
        if data_type == DataType.BIT.value:
            if length % 8 != 0:
                return (length / 8) + 1
            return (length / 8)
        # Use integer size for default
        return length * 2
    
    def get_data_length(self):
        """
        Safely determines the data length for a given message.

        Args:
            None.
        Returns:
            int: The calculated data length.
        Raises:
            None.
        """
        length = 1
        if self.length is not None:
            length = self.length
        return int(length)
    

File: Go/pkg/component/message.go
package component

/*
	Package 'component' provides definitions for various objects meeting the definition of a protocol component.
*/

import (
	"fmt"
	"log/slog"
	"strings"
)

/*
Objects of type Message represent a specific message within a protocol.

Required fields: name.
Other fields are stored in a dictionary corresponding with protocol elements.
*/
type Message struct {
	Name          string                 `json:"name"`
	Message       []byte                 `json:"message"`
	MsgDictionary map[string]interface{} `json:"-"`
}

const (
	Int16  = "int16"
	Int32  = "int32"
	Float  = "float"
	String = "string"
	Bit    = "bit"
)

type DataType string

// NewMessage creates a new instance of Message.
func NewMessage(dataMap map[string]interface{}) *Message {
	message := &Message{}
	if name, ok := dataMap["name"].(string); ok {
		message.Name = name
	} else {
		slog.Error("Message: name field is required")
		return nil
	}
	message.Message = make([]byte, 0)
	// Populate MsgDictionary with remaining key-value pairs
	message.MsgDictionary = make(map[string]interface{})
	for key, value := range dataMap {
		if key != "name" {
			slog.Debug("Adding to MsgDictionary: " + key)
			message.MsgDictionary[key] = value
		}
	}
	message.Log()
	return message
}

func (m *Message) Log() {
	slog.Debug("Message Name: " + m.Name)
	for key, value := range m.MsgDictionary {
		slog.Debug("Message Dictionary Item: " + key + " Value: " + fmt.Sprintf("%v", value))
	}
}

// SetMessage sets the message byte array, which will be used for the duration of the message.
func (m *Message) SetMessage(msg []byte) {
	if len(m.Message) == 0 && msg != nil {
		m.Message = msg
	}
}

// getDataByteCount safely calculates the byte count for a given message.
func (m *Message) getDataByteCount() int {
	length := m.getDataLength()
	dataType := DataType(Int16)
	if m.MsgDictionary["data_type"] != nil {
		dataType = DataType(m.MsgDictionary["data_type"].(string))
	}
	switch strings.ToLower(string(dataType)) {
	case string(DataType(Int16)):
		return int(length) * 2
	case string(DataType(Int32)):
		return int(length) * 4
	case string(DataType(Float)):
		return int(length) * 4
	case string(DataType(String)):
		return int(length)
	case string(DataType(Bit)):
		if int(length)%8 != 0 {
			return int(length/8) + 1
		}
		return int(length) / 8
	default:
		// Use integer size for default
		return int(length) * 2
	}
}

// getDataLength safely determines the length of data for a given message.
func (m *Message) getDataLength() int {
	length := 1
	if m.MsgDictionary["length"] != nil {
		length = int(m.MsgDictionary["length"].(float64))
	}
	return length
}
